File: templates/problems/05_binary_search/07_median_two_sorted/solution.go
package median_two_sorted

// solveFindMedianSortedArrays finds the median using binary search on the shorter array. O(log(min(m,n))) time, O(1) space.
func solveFindMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	// Ensure nums1 is the shorter array
	if len(nums1) > len(nums2) {
		nums1, nums2 = nums2, nums1
	}
	m, n := len(nums1), len(nums2)
	lo, hi := 0, m
	for lo <= hi {
		i := (lo + hi) / 2
		j := (m+n+1)/2 - i

		maxLeftA := minInt
		if i > 0 {
			maxLeftA = nums1[i-1]
		}
		minRightA := maxInt
		if i < m {
			minRightA = nums1[i]
		}
		maxLeftB := minInt
		if j > 0 {
			maxLeftB = nums2[j-1]
		}
		minRightB := maxInt
		if j < n {
			minRightB = nums2[j]
		}

		if maxLeftA <= minRightB && maxLeftB <= minRightA {
			if (m+n)%2 == 1 {
				return float64(max(maxLeftA, maxLeftB))
			}
			return float64(max(maxLeftA, maxLeftB)+min(minRightA, minRightB)) / 2.0
		} else if maxLeftA > minRightB {
			hi = i - 1
		} else {
			lo = i + 1
		}
	}
	return 0.0
}

const (
	minInt = -1000001
	maxInt = 1000001
)

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: templates/problems/05_binary_search/07_median_two_sorted/typescript/solution.ts
export function solveFindMedianSortedArrays(
  nums1: number[],
  nums2: number[],
): number {
  // Find the median using binary search on the shorter array.
  // Time: O(log(min(m, n))), Space: O(1).
  // Ensure nums1 is the shorter array
  if (nums1.length > nums2.length) {
    [nums1, nums2] = [nums2, nums1];
  }
  const m = nums1.length;
  const n = nums2.length;
  let lo = 0;
  let hi = m;
  const inf = 10 ** 9;
  while (lo <= hi) {
    const i = Math.floor((lo + hi) / 2);
    const j = Math.floor((m + n + 1) / 2) - i;

    const maxLeftA = i === 0 ? -inf : nums1[i - 1];
    const minRightA = i === m ? inf : nums1[i];
    const maxLeftB = j === 0 ? -inf : nums2[j - 1];
    const minRightB = j === n ? inf : nums2[j];

    if (maxLeftA <= minRightB && maxLeftB <= minRightA) {
      if ((m + n) % 2 === 1) {
        return Math.max(maxLeftA, maxLeftB);
      }
      return (Math.max(maxLeftA, maxLeftB) + Math.min(minRightA, minRightB)) / 2.0;
    } else if (maxLeftA > minRightB) {
      hi = i - 1;
    } else {
      lo = i + 1;
    }
  }
  return 0.0;
}
